File: projectforge-business/src/main/java/org/projectforge/business/task/ScriptingTaskTree.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.task;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import org.projectforge.business.fibu.AuftragsPositionVO;
import org.projectforge.business.fibu.ProjektDO;


/**
 * Proxy of TaskTree for scripting.
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
public class ScriptingTaskTree
{
  TaskTree __baseDao;

  static List<ScriptingTaskNode> convert(final List<TaskNode> list)
  {
    if (list == null) {
      return null;
    }
    final List<ScriptingTaskNode> result = new ArrayList<ScriptingTaskNode>(list.size());
    for (final TaskNode node : list) {
      result.add(new ScriptingTaskNode(node));
    }
    return result;
  }

  static ScriptingTaskNode convert(final TaskNode node)
  {
    if (node == null) {
      return null;
    }
    return new ScriptingTaskNode(node);
  }

  public ScriptingTaskTree(final TaskTree taskTree)
  {
    this.__baseDao = taskTree;
  }

  public ScriptingTaskNode getRootTaskNode()
  {
    return new ScriptingTaskNode(__baseDao.getRootTaskNode());
  }

  public List<ScriptingTaskNode> getPath(Integer taskId, Integer ancestorTaskId)
  {
    return convert(__baseDao.getPath(taskId, ancestorTaskId));
  }

  public List<ScriptingTaskNode> getPathToRoot(Integer taskId)
  {
    return getPath(taskId, null);
  }

  public ScriptingTaskNode getTaskNodeById(final Integer id)
  {
    return convert(__baseDao.getTaskNodeById(id));
  }

  /**
   * Gets a copy of the found projekt.
   * @param taskId
   * @return
   */
  public ProjektDO getProjekt(final Integer taskId)
  {
    final ProjektDO projekt = __baseDao.getProjekt(taskId);
    if (projekt == null) {
      return null;
    }
    final ProjektDO result = new ProjektDO();
    result.copyValuesFrom(projekt);
    return result;
  }

  public ProjektDO getProjekt(final ScriptingTaskNode node)
  {
    return getProjekt(node.getId());
  }

  public boolean isRootNode(final ScriptingTaskNode node)
  {
    return __baseDao.isRootNode(node.__baseObject);
  }

  public boolean isRootNode(final TaskDO task)
  {
    return __baseDao.isRootNode(task);
  }

  public boolean hasOrderPositionsEntries()
  {
    return __baseDao.hasOrderPositionsEntries();
  }

  public boolean hasOrderPositions(final Integer taskId, final boolean recursive)
  {
    return __baseDao.hasOrderPositions(taskId, recursive);
  }

  public boolean hasOrderPositionsUpwards(final Integer taskId)
  {
    return __baseDao.hasOrderPositionsUpwards(taskId);
  }

  public Set<AuftragsPositionVO> getOrderPositionsUpwards(final Integer taskId)
  {
    return __baseDao.getOrderPositionsUpwards(taskId);
  }

  public BigDecimal getPersonDays(final Integer taskId)
  {
    return __baseDao.getPersonDays(taskId);
  }

  public BigDecimal getPersonDays(final ScriptingTaskNode node)
  {
    return __baseDao.getPersonDays(node.__baseObject);
  }

  public BigDecimal getOrderedPersonDaysSum(final ScriptingTaskNode node)
  {
    return __baseDao.getOrderedPersonDaysSum(node.__baseObject);
  }

  public TaskNode getPersonDaysNode(final ScriptingTaskNode node)
  {
    return __baseDao.getPersonDaysNode(node.__baseObject);
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/business/task/ScriptingTaskTree.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2024 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.task

import org.projectforge.business.fibu.AuftragsPositionVO
import org.projectforge.business.fibu.ProjektDO
import org.projectforge.business.timesheet.TimesheetDO
import java.math.BigDecimal

/**
 * Proxy of TaskTree for scripting.
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
class ScriptingTaskTree {
  private val __baseDao: TaskTree = TaskTree.instance

  val rootTaskNode: ScriptingTaskNode
    get() = ScriptingTaskNode(__baseDao.rootTaskNode)

  fun getPath(taskId: Long?, ancestorTaskId: Long?): List<ScriptingTaskNode>? {
    return convert(__baseDao.getPath(taskId, ancestorTaskId))
  }

  fun getPathToRoot(taskId: Long?): List<ScriptingTaskNode>? {
    return getPath(taskId, null)
  }

  fun getTaskNodeById(id: Long?): ScriptingTaskNode? {
    return convert(__baseDao.getTaskNodeById(id))
  }

  /**
   * Return true, if the given task or any ancestor task matches the given marker.
   * A task matches, if the title, shortDescription, reference or description of a task
   * contains the given marker.
   * @param taskId Check this task for matching and any ancestor task.
   * @param marker Marker must be part of task or any ancestor task. If null or blank, any task matches.
   */
  @JvmOverloads
  fun matchesMarker(taskId: Long?, marker: String?, ignoreCase: Boolean = true): Boolean {
    var taskNode: TaskNode? = __baseDao.getTaskNodeById(taskId) ?: return false
    if (marker.isNullOrBlank()) {
      return true
    }
    for (i in 0..1000) { // For loop as paranoia loop for avoiding circular references in task nodes.
      val task = taskNode?.task
      if (task != null
        && (task.title?.contains(marker, ignoreCase = ignoreCase) == true
            || task.shortDescription?.contains(marker, ignoreCase = ignoreCase) == true
            || task.reference?.contains(marker, ignoreCase = ignoreCase) == true
            || task.description?.contains(marker, ignoreCase = ignoreCase) == true)
      ) {
        return true
      }
      taskNode = taskNode?.parent
    }
    return false
  }

  class FilteredTimesheets {
    val filteredTimesheets = mutableListOf<TimesheetDO>()
    val matchingTasks = mutableSetOf<TaskDO>()
    val notMatchingTasks = mutableSetOf<TaskDO>()
  }

  /**
   * Filter the timesheets by matching tasks.
   * @see matchesMarker
   */
  fun filterTimesheets(timesheets: List<TimesheetDO>, marker: String?, ignoreCase: Boolean = true): FilteredTimesheets {
    val result = FilteredTimesheets()
    timesheets.forEach { timesheet ->
      val task = timesheet.task
      if (task != null) {
        if (matchesMarker(task.id, marker, ignoreCase)) {
          if (!result.matchingTasks.any { it.id == task.id }) {
            result.matchingTasks.add(task)
          }
          result.filteredTimesheets.add(timesheet)
        } else {
          if (!result.notMatchingTasks.any { it.id == task.id }) {
            result.notMatchingTasks.add(task)
          }
        }
      }
    }
    return result
  }

  /**
   * Gets a copy of the found projekt.
   * @param taskId
   * @return
   */
  fun getProjekt(taskId: Long?): ProjektDO? {
    val projekt = __baseDao.getProjekt(taskId) ?: return null
    val result = ProjektDO()
    result.copyValuesFrom(projekt)
    return result
  }

  fun getProjekt(node: ScriptingTaskNode): ProjektDO? {
    return getProjekt(node.id)
  }

  fun isRootNode(node: ScriptingTaskNode): Boolean {
    return __baseDao.isRootNode(node.__baseObject)
  }

  fun isRootNode(task: TaskDO): Boolean {
    return __baseDao.isRootNode(task)
  }

  fun hasOrderPositionsEntries(): Boolean {
    return __baseDao.hasOrderPositionsEntries()
  }

  fun hasOrderPositions(taskId: Long?, recursive: Boolean): Boolean {
    return __baseDao.hasOrderPositions(taskId, recursive)
  }

  fun hasOrderPositionsUpwards(taskId: Long?): Boolean {
    return __baseDao.hasOrderPositionsUpwards(taskId)
  }

  fun getOrderPositionsUpwards(taskId: Long?): Set<AuftragsPositionVO> {
    return __baseDao.getOrderPositionsUpwards(taskId)
  }

  fun getPersonDays(taskId: Long?): BigDecimal {
    taskId ?: return BigDecimal.ZERO
    return __baseDao.getPersonDays(taskId) ?: BigDecimal.ZERO
  }

  fun getPersonDays(node: ScriptingTaskNode?): BigDecimal {
    if (node?.__baseObject == null) {
      return BigDecimal.ZERO
    }
    return __baseDao.getPersonDays(node.__baseObject) ?: BigDecimal.ZERO
  }

  fun getOrderedPersonDaysSum(node: ScriptingTaskNode?): BigDecimal {
    if (node?.__baseObject == null) {
      return BigDecimal.ZERO
    }
    return __baseDao.getOrderedPersonDaysSum(node.__baseObject) ?: BigDecimal.ZERO
  }

  fun getPersonDaysNode(node: ScriptingTaskNode): TaskNode? {
    return __baseDao.getPersonDaysNode(node.__baseObject)
  }

  companion object {
    @JvmStatic
    fun convert(list: List<TaskNode?>?): List<ScriptingTaskNode>? {
      if (list == null) {
        return null
      }
      val result: MutableList<ScriptingTaskNode> = ArrayList(list.size)
      for (node in list) {
        result.add(ScriptingTaskNode(node))
      }
      return result
    }

    @JvmStatic
    fun convert(node: TaskNode?): ScriptingTaskNode? {
      return node?.let { ScriptingTaskNode(it) }
    }
  }
}
